File: src/main/java/at/searles/parsing/combinators/RecognizerOpt.java
package at.searles.parsing.combinators;

import at.searles.parsing.Recognizable;
import at.searles.parsing.Recognizer;
import at.searles.parsing.printing.ConcreteSyntaxTree;
import org.jetbrains.annotations.NotNull;

/**
 * Optional recognizer. This one is different from reducer and parser
 * because its printer is not semantically equivalent to A | epsilon.
 */
public class RecognizerOpt implements Recognizer, Recognizable.Opt {

    private final Recognizer parent;
    private final boolean alwaysPrint;

    public RecognizerOpt(Recognizer parent, boolean alwaysPrint) {
        this.parent = parent;
        this.alwaysPrint = alwaysPrint;
    }

    @Override
    public Recognizable parent() {
        return parent;
    }

    @NotNull
    @Override
    public ConcreteSyntaxTree print() {
        if (alwaysPrint) {
            return parent.print();
        }

        return ConcreteSyntaxTree.empty();
    }

    @Override
    public String toString() {
        return createString();
    }
}

File: src/main/java/at/searles/parsing/combinators/RecognizerOpt.kt
package at.searles.parsing.combinators

import at.searles.parsing.ParserStream
import at.searles.parsing.Recognizer
import at.searles.parsing.printing.ConcreteSyntaxTree
import at.searles.parsing.printing.EmptyTree

/**
 * Optional recognizer. This one is different from reducer and parser
 * because its printer is not semantically equivalent to A | epsilon.
 */
class RecognizerOpt(private val parent: Recognizer, private val alwaysPrint: Boolean = false) : Recognizer {
    override fun recognize(stream: ParserStream): Boolean {
        stream.recognize(parent, true)
        return true
    }

    override fun print(): ConcreteSyntaxTree {
        return if(alwaysPrint) parent.print() else EmptyTree
    }

    override fun toString(): String {
        return "$parent.opt"
    }
}